crates/vm/src/arch/execution_mode/metered_cost.rs

```diff
@@ -0,0 +1,140 @@
+use std::num::NonZero;
+
+use openvm_instructions::riscv::RV32_IMM_AS;
+
+use crate::{
+    arch::{ExecutionCtxTrait, MeteredExecutionCtxTrait, VmExecState, PUBLIC_VALUES_AIR_ID},
+    system::memory::online::GuestMemory,
+};
+
+#[derive(Clone, Debug)]
+pub struct AccessAdapterCtx {
+    as_byte_alignment_bits: Vec<u8>,
+    idx_offset: usize,
+}
+
+impl AccessAdapterCtx {
+    pub fn new(
+        as_byte_alignment_bits: Vec<u8>,
+        has_public_values_chip: bool,
+        continuations_enabled: bool,
+    ) -> Self {
+        let boundary_idx = if has_public_values_chip {
```

Comment:
I do wish we could remove the duplication of this special logic (it's in various places unfortunately)